demucs/istft.py

```diff
@@ -0,0 +1,164 @@
+        super(ISTFT_Process, self).__init__()
+        self.n_fft = n_fft
+        self.hop_len = hop_len
+        self.window_type = window_type
+        self.window_length = window_length
+        self.normalized = normalized
+        self.max_frames = max_frames
+        self.center = center
+        self.half_n_fft = n_fft // 2  # Precompute once
+
+        # Get window function and compute window once
+        if self.window_length != self.n_fft:
+            raise NotImplementedError(f"The case of window length not equal to n_fft is not implemented in {self.__class__.__name__}.")
+        window = WindowType(window_type)(self.window_length).float()
+
+        # Check if center is false
+        if not self.center:
+            raise NotImplementedError("No centering is not supported in this implementation.")
+
+        # ISTFT forward pass preparation
+        # Pre-compute fourier basis
+        fourier_basis = torch.fft.fft(torch.eye(n_fft, dtype=torch.float32))
+        fourier_basis = torch.vstack([
+            torch.real(fourier_basis[:self.half_n_fft + 1, :]),
+            torch.imag(fourier_basis[:self.half_n_fft + 1, :])
+        ]).float()
+        
+        # Create forward and inverse basis
+        forward_basis = window * fourier_basis[:, None, :]
+        inverse_basis = window * torch.linalg.pinv((fourier_basis * n_fft) / hop_len).T[:, None, :]
+        
+        # Calculate window sum for overlap-add
+        n = n_fft + hop_len * (max_frames - 1)
+        window_sum = torch.zeros(n, dtype=torch.float32)
+        window_normalized = window / window.abs().max()
+        
+        # Pad window if needed
+        total_pad = n_fft - window_normalized.shape[0]
+        if total_pad > 0:
+            pad_left = total_pad // 2
+            pad_right = total_pad - pad_left
+            win_sq = torch.nn.functional.pad(window_normalized ** 2, (pad_left, pad_right), mode='constant', value=0)
+        else:
+            win_sq = window_normalized ** 2
+        
+        # Calculate overlap-add weights
+        for i in range(max_frames):
+            sample = i * hop_len
+            window_sum[sample: min(n, sample + n_fft)] += win_sq[: max(0, min(n_fft, n - sample))]
+        
+        # Normalize window if needed
+        if normalized:
+            inverse_basis = inverse_basis * torch.sqrt(torch.tensor([n_fft], dtype=torch.float32))
+        
+        # Register buffers
+        self.register_buffer("forward_basis", forward_basis)
+        self.register_buffer("inverse_basis", inverse_basis)
+        self.register_buffer("window_sum_inv", n_fft / (window_sum * hop_len + 1e-8))  # Add epsilon to avoid division by zero
+
+    def forward(self, real, imag, length=None):
+        # Calculate magnitude and phase from real and imaginary parts
+        magnitude = torch.sqrt(real ** 2 + imag ** 2)
+        phase = torch.atan2(imag, real + 1e-7)  # Add epsilon to avoid division by zero
```

Comment:
it'd be great if we could use a constant in here, as epsilon could change based of the float precision (f32 != f64). It's also generally a good practice to store these constant outside of your implementation.

Looking at this [torch.finfo](https://docs.pytorch.org/docs/stable/type_info.html#torch.torch.finfo), you might be able to do 

```suggestion
        phase = torch.atan2(imag, real + torch.finfo().eps)  # Add epsilon to avoid division by zero
```